File: challenges/hackerrank/EvenOddQuery.py
"""
A module to solve the problem specified at:
    https://www.hackerrank.com/contests/w5/challenges/even-odd-query

You are given an array A of size N. You are also given an integer Q. Can you
figure out the answer to each of the Q queries?

Each query contains 2 integers x and y, and you need to find whether the value
find(x,y) is odd or even:

find(int x,int y)
{
    if(x>y) return 1;
    ans = pow(A[x],find(x+1,y))
    return ans
}

Sample Input
3
3 2 7
2
1 2
2 3

Sample Output
Odd
Even
"""


# pylint: disable=invalid-name
# pylint: disable=bad-builtin
def find(array, x, y):
    """
    The crux function
    """
    if x > y:
        return 1
    out = array[x] ** (1 if array[x+1] else 0)
    return out


def isOdd(number):
    """
    Returns whether the provided argument is an odd-number i.e. divisible by 2.
    """
    return number % 2 == 1


def getOddOrEvenLabel(numberIsOdd):
    """
    Returns "Odd" if numberIsOdd argument is True; "Even" otherwise.
    """
    return "Odd" if numberIsOdd else "Even"


def main():
    """
    Ugh, pylint, this is for you!
    """
    n = long(raw_input())
    if n <= 0:
        return
    array = map(long, raw_input().split(" "))
    n = long(raw_input())
    if n <= 0:
        return
    pairs = []
    for _ in range(n):
        x, y = raw_input().split(" ")
        pairs.append({'x': long(x), 'y': long(y)})
    for pair in pairs:
        x = pair['x']
        y = pair['y']
        numberIsOdd = isOdd(find(array, x-1, y-1))
        print getOddOrEvenLabel(numberIsOdd)


if __name__ == "__main__":
    main()

File: challenges/hackerrank/EvenOddQuery.cpp
#include "EvenOddQuery.h"

int main() {
  unsigned n;
  cin >> n;
  if (n == 0) {
    return 0;
  }

  vector<unsigned> numbers;
  for(unsigned i = 0; i < n; i++) {
    unsigned number;
    cin >> number;
    numbers.push_back(number);
  }

  unsigned num_pairs;
  cin >> num_pairs;
  if (num_pairs == 0) {
    return 0;
  }

  typedef std::pair<unsigned, unsigned> pair;
  vector<pair> pairs;
  for(unsigned i = 0; i < num_pairs; i++) {
    unsigned x, y;
    cin >> x; cin >> y;
    pairs.push_back(make_pair(x, y));
  }
  for(unsigned i = 0; i < num_pairs; i++) {
    unsigned x = pairs[i].first;
    unsigned y = pairs[i].second;
#if DEBUG
    cout << x << ", " << y << ": " ;
#endif
    bool number_is_odd = isOdd(find(numbers, x-1, y-1));
    string out = number_is_odd ? "Odd" : "Even";
    cout << out << endl;
  }
  return 0;
}

unsigned find(
    vector<unsigned> numbers,
    unsigned x,
    unsigned y) {
#if DEBUG
  cout << "(" << x << ", " << y << "), ";
#endif
  if (x > y) {
#if DEBUG==2
    cout << "Returning 1 for x: " << x << ", y: " << y << " ";
#endif
    return 1;
  }
#if BASE_ONLY
  unsigned base = numbers[x];
  unsigned exponent = numbers[x+1] == 0 ? 0 : 1;
  long double out = pow(base, exponent);
#if DEBUG==2
  cout << "(" << x << ", " << y << "): pow(" << base <<
    ", " << exponent << ") = " << out << ", ";
#endif
  return out;
#else
  unsigned base = numbers[x];
  unsigned exponent = find(numbers, x+1, y);
  long double out = pow(base, exponent);
#if DEBUG==2
  cout << "(" << x << ", " << y << "): pow(" << base <<
    ", " << exponent << ") = " << out << ", ";
#endif
  return out;
#endif
}

bool isOdd(unsigned x) {
  return x % 2;
}
